### PR TITLE
Bug lp:1667307 fix.

### PR DESCRIPTION
**Pull request by vlad-lesin**
Bug description:

1) When binlog_rows_query_log_events is enabled, Rows_query_log_event
is created, written to binary log on master prior Rows_log_event.

2) Slave is reading this event from relay log and applying it,
see Rows_query_log_event::do_apply_event(). When the event is applied
the event itself is stored in Relay_log_info::rows_query_ev:
```c++
int Rows_query_log_event::do_apply_event(Relay_log_info const *rli)
{
  DBUG_ENTER("Rows_query_log_event::do_apply_event");
  DBUG_ASSERT(rli->info_thd == thd);
  /* Set query for writing Rows_query log event into binlog later.*/
  thd->set_query(m_rows_query, strlen(m_rows_query));

  DBUG_ASSERT(rli->rows_query_ev == NULL);

  const_cast<Relay_log_info*>(rli)->rows_query_ev= this;
  /* Tell worker not to free the event */
  worker= NULL;
  DBUG_RETURN(0);
}
```

3) Usually Relay_log_info::rows_query_ev is removed after Rows_log_event
execution with the following stack trace:

```
#0 Rows_query_log_event::~Rows_query_log_event (this=0x7fff9401df00, __in_chrg=<optimized out>, __vtt_parm=<optimized out>)
    at ./sql/log_event.h:3795
#1 0x0000000001958d1e in Relay_log_info::cleanup_context (this=0x7fffa40413d0, thd=0x7fff94000950, error=false)
    at ./sql/rpl_rli.cc:1777
#2 0x00000000018d984b in rows_event_stmt_cleanup (rli=0x7fffa40413d0, thd=0x7fff94000950) at ./sql/log_event.cc:11297
#3 0x00000000018d9500 in Rows_log_event::do_apply_event (this=0x7fff94012fc0, rli=0x7fffa40413d0)
    at ./sql/log_event.cc:11175
#4 0x00000000018c00c7 in Log_event::apply_event (this=0x7fff94012fc0, rli=0x7fffa40413d0) at ./sql/log_event.cc:3371
#5 0x0000000001938e0f in apply_event_and_update_pos (ptr_ev=0x7ffff139e8a0, thd=0x7fff94000950, rli=0x7fffa40413d0)
    at ./sql/rpl_slave.cc:4750
#6 0x000000000193a5d7 in exec_relay_log_event (thd=0x7fff94000950, rli=0x7fffa40413d0) at ./sql/rpl_slave.cc:5265
#7 0x00000000019418eb in handle_slave_sql (arg=0x7fffa4139800) at ./sql/rpl_slave.cc:7461
#8 0x0000000001e6a679 in pfs_spawn_thread (arg=0x7fffa413d640) at ./storage/perfschema/pfs.cc:2188
#9 0x00007ffff71616ba in start_thread (arg=0x7ffff139f700) at pthread_create.c:333
#10 0x00007ffff65f682d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
```

But if we look into Rows_log_event::do_apply_event() the function
rows_event_stmt_cleanup() is called at the end:

```c++
int Rows_log_event::do_apply_event(Relay_log_info const *rli)
{
...
end:
  if (get_flags(STMT_END_F))
  {
    if((error= rows_event_stmt_cleanup(rli, thd)))
    {
      ...
    }
    ...
  }
...
}
```

4) But when the statement must be skipped because it has been already
applied from another channel, Rows_log_event::do_apply_event() exits before
rows_event_stmt_cleanup() call. See the following code:

```c++
int Rows_log_event::do_apply_event(Relay_log_info const *rli)
{
...
    enum_gtid_statement_status state= gtid_pre_statement_checks(thd);
    if (state == GTID_STATEMENT_EXECUTE)
    {
      if (gtid_pre_statement_post_implicit_commit_checks(thd))
        state= GTID_STATEMENT_CANCEL;
    }

    if (state == GTID_STATEMENT_CANCEL)
    {
      uint error= thd->get_stmt_da()->mysql_errno();
      DBUG_ASSERT(error != 0);
      rli->report(ERROR_LEVEL, error,
                  "Error executing row event: '%s'",
                  thd->get_stmt_da()->message_text());
      thd->is_slave_error= 1;
      DBUG_RETURN(-1);
    }
    else if (state == GTID_STATEMENT_SKIP)
      DBUG_RETURN(0);
...
end:
  if (get_flags(STMT_END_F))
  {
    if((error= rows_event_stmt_cleanup(rli, thd)))
    {
      ...
    }
    ...
  }
...
}
```

So in this case rli->rows_query_ev will not be deleted. In debug build such
case will lead to crash due to the corresponding assert in
Rows_query_log_event::do_apply_event():

```c++
int Rows_query_log_event::do_apply_event(Relay_log_info const *rli)
{
...
  DBUG_ASSERT(rli->rows_query_ev == NULL);
...
}
```

But for release build there will be just memory leak.

How to fix:

delete Rows_query_log_event instance if the next Rows_log_event instance read
from relay log must be skipped.

Testing: http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/983/
rpl_multi_source_slave_files looks very suspicious but it fails without my fix, and here is the bug report: https://bugs.mysql.com/bug.php?id=79402
See also: https://bugs.launchpad.net/percona-server/+bug/1667307

### CODE REVIEW

File: mysql-test/suite/rpl/r/rpl_bug-lp-1667307.result
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/r/rpl_bug-lp-1667307.result]: Since this is an upstream bug, use upstream bug number rpl.bug85371 for testcase naming (for easier upstream submission - see Confluence)

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: mysql-test/suite/rpl/t/rpl_bug-lp-1667307.test
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/rpl_bug-lp-1667307.test]: Also `--source include/not_group_replication_plugin.inc`

File: mysql-test/suite/rpl/t/rpl_bug-lp-1667307.test
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/rpl_bug-lp-1667307.test]: Why do we need `have_debug.inc`?

File: mysql-test/suite/rpl/t/rpl_bug-lp-1667307.cnf
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/rpl_bug-lp-1667307.cnf]: Strictly speaking, we don't need
`master-info-repository=TABLE`
`relay-log-info-repository=TABLE`
for the first server

File: mysql-test/suite/rpl/t/rpl_bug-lp-1667307.test
Comment (percona-ysorokin) [inline mysql-test/suite/rpl/t/rpl_bug-lp-1667307.test]: test case naming as per Laurynas' comments

File: sql/log_event.cc
Comment (percona-ysorokin) [inline sql/log_event.cc:10726]: Do we need similar cleanup logic for the previous branch
`if (state == GTID_STATEMENT_CANCEL)`?

**Review by percona-ysorokin — CHANGES_REQUESTED** [review]

File: mysql-test/suite/rpl/r/rpl_bug-lp-1667307.result
Comment (vlad-lesin) [reply mysql-test/suite/rpl/r/rpl_bug-lp-1667307.result]: Fixed.

File: mysql-test/suite/rpl/t/rpl_bug-lp-1667307.test
Comment (vlad-lesin) [reply mysql-test/suite/rpl/t/rpl_bug-lp-1667307.test]: Fixed.

File: mysql-test/suite/rpl/t/rpl_bug-lp-1667307.test
Comment (vlad-lesin) [reply mysql-test/suite/rpl/t/rpl_bug-lp-1667307.test]: Fixed.

File: mysql-test/suite/rpl/t/rpl_bug-lp-1667307.test
Comment (vlad-lesin) [reply mysql-test/suite/rpl/t/rpl_bug-lp-1667307.test]: Yes, we need it, because the "assert", on which the test crashes, if the bug is not fixed, happens only on debug build as "DBUG_ASSERT" macros is used. See the following code:

```c++
int Rows_query_log_event::do_apply_event(Relay_log_info const *rli)
{
  ...
  DBUG_ASSERT(rli->rows_query_ev == NULL);
...
}
```

File: mysql-test/suite/rpl/t/rpl_bug-lp-1667307.cnf
Comment (vlad-lesin) [reply mysql-test/suite/rpl/t/rpl_bug-lp-1667307.cnf]: No, we need it, see topology in the test source code:

```
--let $rpl_topology= 1->3,3->2,1->2
--let $rpl_multi_source= 1
--let $use_gtids= 1
--source include/rpl_init.inc
```

For such topology we need multi-source replication, what requires master and relay log info format to be "TABLE". I am not sure if it is possible to have some servers with "multisource" option on, and some of them with the option switched off in replication graph, but, anyway, "rpl_init.inc" does not allow this, if "$rpl_multi_source" is on, then it's on for all servers pointed in the graph.

File: sql/log_event.cc
Comment (vlad-lesin) [reply sql/log_event.cc:10726]: I am not sure about this. This question requires additional research and there indeed must be other test to model different situations(I counted at least four such cases) when gtid_pre_statement_checks() returns GTID_STATEMENT_CANCEL and how higher-level functions process the case when Rows_log_event::do_apply_event() returns error. I think such cases must be considered separately from the original case, and there must be separate bug report.

File: mysql-test/suite/rpl/t/rpl_bug-lp-1667307.test
Comment (laurynas-biveinis) [reply mysql-test/suite/rpl/t/rpl_bug-lp-1667307.test]: I see no reason not to run this on release build as well, this a regression test case after all

**Comment by percona-ysorokin** [comment]
Looks good with the only comment that I also think that this test case must be run not only in debug mode (even though the initial problem was caused by `DBUG_ASSERT`, it won't hurt to enable this test case in release mode anyway).

**Comment by vlad-lesin** [comment]
Yura, including this test in release testing does not have any sense because for release build it will pass anyway, regardless the fact of regression.

And the initial problem was caused by memory leak, not by DBUG_ASSERT.

File: mysql-test/suite/rpl/t/rpl_bug-lp-1667307.test
Comment (vlad-lesin) [reply mysql-test/suite/rpl/t/rpl_bug-lp-1667307.test]: Laurynas, I just want to clarify about regression testing. My understanding of regression supposes that there are the criteria of regression. And this is not only my understanding. https://en.wikipedia.org/wiki/Regression_testing : "Regression testing is a type of software testing which verifies that software which was previously developed and tested still performs correctly after it was changed or interfaced with other software." What means the criteria of correctness must exist. In our case, such criteria exist only in debug build. Even if something is broken in release build we will not see this in the current version of the test. In this test, it is supposed that some code (assertion) must be executed to specify such criteria, what happens only in debug build. So what the sense of this test in a release build if there is no way to catch regression for release build?

Comment (laurynas-biveinis) [reply mysql-test/suite/rpl/t/rpl_bug-lp-1667307.test]: For the release build the criteria would be admittedly weaker "runs, does not crash, completes correctly". According to your criteria the testcase should contain e.g. --source include/not_valgrind.inc, not_asan.inc at least, as it only trips on DBUG_ASSERT, shouldn't it? Having said that, if you considered everything above and still think this must be a debug-only testcase, then I'm OK with your decision. (I don't know about other reviewers of course)

Comment (vlad-lesin) [reply mysql-test/suite/rpl/t/rpl_bug-lp-1667307.test]: Good point, Laurynas. valgrind should catch it. "have_debug.inc" has been removed.

**Comment by vlad-lesin** [comment]
Guys, it seems there is no more dissidence for this pull request. I am waiting for your approvals to merge it.

**Review by percona-ysorokin — APPROVED** [review]
LGTM

**Review by laurynas-biveinis — APPROVED** [review]
My review is only partial

**Pull request merged**
